### ISSUE TITLE
Make lazy-loader compilation more automatic

### ISSUE THREAD

**Issue by wbthomason** [wbthomason/packer.nvim#7]
As mentioned in #4, it's inconvenient and confusing to have to call `packer.compile()` manually. This process should be made more automatic.

One option is to add a config key containing a path for the compiler output. If this key exists, `compile()` gets run automatically after all operations.

Additionally (as @tjdevries suggested on Gitter), perhaps the lazy-loader could check if the last modified time of the plugin specification file is after the last modified time of the compiled lazy-loaders, and update itself if so.

**Comment by tjdevries** [comment]
I think packer should formalize the "lua plugins" file in some way (probably by having some configuration file).

For example: `packer.init(plug_file)` or similar, and do some of the setup and command creation that most other package managers use.

**Comment by wbthomason** [comment]
See #10 for part of this.

**Comment by wbthomason** [comment]
Hmm, the problem I'm running into with this at the moment is that the loader needs to know what code to run to initialize the set of managed plugins in order to recompile itself. This is one place that `packer`'s flexibility hurts it - that code could be a named function, a file, or a closure (so we might not even have access to it)...

**Comment by tjdevries** [comment]
Yeah, you could provide a hook for determining if compilation is necessary and then `startup` could have a default hook ( like... idk? maybe something though :laughing: ) but at least with that one we have a specific way to set it up or not.

For example, you could save in `stdpath('cache')` a loadable string of the contents of `plugin` and compare that to the current `plugins` or similar.

**Comment by wbthomason** [comment]
So, determining if the loaders are out of date (to an approximation) is pretty easy - `stat` the loaders file and the specification file, compare `mtime` values. The hard part is that - in the context of the compiled loaders - the table of plugin specifications generally has not been loaded, and it's hard to know how to generate that for the aforementioned reason.

A much easier improvement would be to just run `packer.compile()` at the end of e.g. `packer.sync()`, but that doesn't capture most of the cases in which I want to rebuild the loaders.

**Comment by Shougo** [comment]
> As mentioned in #4, it's inconvenient and confusing to have to call packer.compile() manually. This process should be made more automatic.

Hm.  But users can define autocommands for it.

`autocmd BufWritePost init.vim ...`

It can be automatically in packer.nvim, but it seems be complex and may have bug.

**Comment by MaxCan-Code** [comment]
The straight package manager has the option to use [`watchexec`](https://github.com/watchexec/watchexec) to recompile changed files automatically: https://github.com/raxod502/straight.el#my-init-time-got-slower

Running `watchexec` in a background subshell like
```sh
{watchexec -pw ~/.config/nvim/init.lua 'nvim +"au User PackerComplete qa" +PackerSync' &}
```
will run `PackerSync` whenever there's a change written to `~/.config/nvim/init.lua`

I run that on a separate terminal session to avoid the PackerSync popup
